libraries/ESP8266WebServer/src/ESP8266WebServerSecure.h

```diff
@@ -0,0 +1,55 @@
+/*
+  ESP8266WebServerSecure.h - Dead simple HTTPS web-server.
+  Supports only one simultaneous client, knows how to handle GET and POST.
+
+  Copyright (c) 2017 Earle F. Philhower, III. All rights reserved.
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+
+#ifndef ESP8266WEBSERVERSECURE_H
+#define ESP8266WEBSERVERSECURE_H
+
+#include <ESP8266WebServer.h>
+#include <WiFiServerSecure.h>
+
+class ESP8266WebServerSecure : public ESP8266WebServer
+{
+public:
+  ESP8266WebServerSecure(IPAddress addr, int port = 443);
+  ESP8266WebServerSecure(int port = 443);
+  ~ESP8266WebServerSecure();
+
+  void setServerKeyAndCert_P(const uint8_t *key, int keyLen, const uint8_t *cert, int certLen);
+  void setServerKeyAndCert(const uint8_t *key, int keyLen, const uint8_t *cert, int certLen);
+
+  WiFiClient client() override { return _currentClientSecure; }
+
+  void begin() override;
+  void handleClient() override;
+  void close() override;
+
+private:
+  size_t _currentClientWrite (const char *bytes, size_t len) override { return _currentClientSecure.write((const uint8_t *)bytes, len); }
+  size_t _currentClientWrite_P (PGM_P bytes, size_t len) override { return _currentClientSecure.write_P(bytes, len); }
+
+protected:
+  WiFiServerSecure _serverSecure;
+  WiFiClientSecure _currentClientSecure;
```

Comment:
Hmm, so ESP8266WebServerSecure will still carry _server and _currentClient members in addition to these new ones. Perhaps we can refactor this by turning ESP8266WebServer into a template, like `ESP8266WebServerImpl<typename TServer, typename TClient>` and then doing a `typedef ESP8266WebServerImpl<WiFiServer, WiFiClient> ESP8266WebServer` and `typedef ESP8266WebServerImpl<WiFiServerSecure, WiFiClientSecure> ESP8266WebServerSecure`.
Small part of the implementation, of course, will need to be moved into the header file.

(I'm okay with doing this refactoring in a subsequent PR)

Reply:
You're right, it is crufty.  Can we do this on a later PR?  I've used templates, of course, but never had to write my own.  It's probably not as bad as I imagine, but it's something I'd need to spend a bit reading up on.

Another alternative might be to convert to pass-by-reference instead of -by-value.  Then it may "just work" without worrying if it's a WiFiServer or WiFiServerSecure in the methods that don't care about SSL.  That's probably a larger change, though, and adds dynamic memory allocations (but String is used everywhere so I'd say that worry is not a big deal).

Reply:
Every time I see inheritance and virtual etc in an embedded project I cry a little. Inheritance is really ill suited for systems that are really memory constrained. There are better ways.
Having said that, I don't mind handling that change separately. I've been sort of thinking about an overhaul of some classes into a non-inheritance implementation anyways, but that would be a Big Change, and right now I'd like to stay focused on stability and cleanup.

Reply:
Yes, I'm okay doing this in a later PR. Just thought I'd mention this here. Passing things by reference will indeed help with code duplication.